docs/manual/src/udl/decorators.md

```diff
@@ -0,0 +1,240 @@
+
+A common pattern when using uniffied code is to run some common, but app-specific, code before or after calling the Rust code.
+
+```kotlin
+class DemoLib(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+    val listener: () -> Void
+) {
+    val demoRustObject = new DemoRustObject()
+
+    fun doExpensiveThing() {
+        backgroundScope.launch {
+            demoRustObject.doExpensiveThing()
+        }
+    }
+
+    fun doCrashyThing() {
+        try {
+            demoRustObject.doCrashyThing()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+    }
+
+    fun doConsequentialThing() {
+        demoRustObject.doConsequentialThing()
+        listener()
+    }
+}
+```
+
+This causes a proliferation of boiler plate code. Worse, everytime we add a new method to the `DemoRustObject`, handwritten foreign language code needs to be written to expose it safely.
+
+The more methods that want to do similar things, the more repetitive this gets.
+
+We could isolate the repeated code into a decorator class with `onBackgroundThread`, `withErrorReporter` and `thenNotifyListeners` methods.
+
+We'll handwrite a little decorator interface, then write an implementation that we pass the thread pool/coroutine context/whatever, and the error reporter.
+
+```kotlin
+interface DemoDecorator {
+    fun onBackgroundThread(rustCall: () -> Unit)
+    fun <T> withErrorReporter(rustCall: () -> T): T
+    fun thenNotifyListeners(obj: DemoLib, rustCall: () -> T)
+}
+
+class MyDemoDecorator(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+    val listener: () -> Void
+) : DemoDecorator {
+    fun onBackgroundThread(rustCall: () -> Unit) {
+        backgroundScope.launch {
+            rustCall()
+        }
+    }
+
+    fun <T> withErrorReporter(rustCall: () -> T) =
+        try {
+            rustCall()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun thenNotifyListeners(obj: DemoLib, rustCall: () -> T) {
+        try {
+            rustCall()
+        } finally {
+            listener()
+        }
+    }
+}
+```
+
+Then, we can re-write the `DemoLib` as:
+
+```kotlin
+class DemoLib(
+    val decorator: DemoDecorator = MyDemoDecorator()
+) {
+    val demoRustObject = new DemoRustObject()
+
+    fun doExpensiveThing() = decorator.onBackgroundThread {
+        demoRustObject.doExpensiveThing()
+    }
+
+    fun doCrashyThing() = decorator.withErrorReporter {
+        demoRustObject.doCrashyThing()
+    }
+
+    fun doConsequentialThing() = decorator.thenNotifyListeners(this) {
+        demoRustObject.doConsequentialThing()
+    }
+}
+```
+
+This is much better, but it's still looking a bit cut and pasty. Enter decorator objects.
+
+## Solution
+
+```webidl
+namespace demo {}
+
+[Decorator]
+interface DemoDecorator {
+    Any? with_error_reporter();
+    void on_background_thread();
+    Any  then_notify_listeners();
+}
```

Comment:
I think this should be in `uniffi.toml` and not the UDL.  This gives a couple benefits:

 - It allows consumers to use different return types.  A different consumer might want `with_error_reporter` to log exceptions, but still propagate them up the stack.
 - It allows support for return types that UniFFI doesn't know about and/or types that are spelled differently in different languages.  The main example here are async types.  I completely agree that UniFFI shouldn't force users into any particular async framework, which probably means that UniFFI can't know what the return type should be.  Also, each language has slightly different types names.  It seems like the best way to handle this is to allow consumers to specify the return type in `uniffi.toml`.  One consumer could use `Deferred<T>`, another could use `Promise<T>`, etc.  I see a lot of parallels between this and wrapped types.
 - We don't need to define the `DecoratorObject` and `Generic` types.  I don't like adding 2 new types that work very differently than all the others.  This is more of a minor issue, but something about it feels off to me.

However, I think there's still some disagreement on this one and even if there wasn't we would need to figure out what the syntax would be.  In the meantime, could we defer the decision by just leaving this interface definition out?  It seems like we could implement a first-cut of the feature without this information:
  - The interface name is specified as part of the `[Decorator]` attribute
  - The method names are specified in the `[CallsWith]` attributes
  - Kotlin can infer the return types.  Python and Ruby don't need them.  Swift may or may not be able to do the inference, but this PR only targets Kotlin to start.